src/__tests__/interception.test.ts

```diff
@@ -0,0 +1,285 @@
+import {
+  instance as globalContainer,
+  ResolutionType
+} from "../dependency-container";
+import Frequency from "../types/frequency";
+
+// beforeResolution .resolve() tests
+test("beforeResolution interceptor gets called correctly", () => {
+  class Bar {}
+  let interceptorCalled = false;
+  globalContainer.beforeResolution(Bar, (_, resolutionType) => {
+    interceptorCalled = true;
+    expect(resolutionType).toEqual(ResolutionType.Single);
+  });
+  globalContainer.resolve(Bar);
+
+  expect(interceptorCalled).toBeTruthy();
+});
+test("beforeResolution interceptor gets called correctly", () => {
+  class Bar {}
+  let interceptorCalled = false;
+  globalContainer.beforeResolution(Bar, () => {
+    interceptorCalled = true;
+  });
+  globalContainer.resolve(Bar);
+
+  expect(interceptorCalled).toBeTruthy();
+});
+
+test("beforeResolution interceptor does not get called when resolving other types", () => {
+  class Bar {}
+  class Foo {}
+  let interceptorCalled = false;
+  globalContainer.beforeResolution(Bar, () => {
+    interceptorCalled = true;
+  });
+  globalContainer.resolve(Foo);
+
+  expect(interceptorCalled).toBeFalsy();
+});
+
+test("beforeResolution one-time interceptor only gets called once", () => {
+  class Bar {}
+  let timesCalled = 0;
+  globalContainer.beforeResolution(
+    Bar,
+    () => {
+      timesCalled++;
+    },
+    {frequency: Frequency.Once}
+  );
+  globalContainer.resolve(Bar);
+  globalContainer.resolve(Bar);
+
+  expect(timesCalled).toEqual(1);
+});
+
+test("beforeResolution always run interceptor gets called on each resolution", () => {
+  class Bar {}
+  let timesCalled = 0;
+  globalContainer.beforeResolution(
+    Bar,
+    () => {
+      timesCalled++;
+    },
+    {frequency: Frequency.Always}
+  );
+  globalContainer.resolve(Bar);
+  globalContainer.resolve(Bar);
+
+  expect(timesCalled).toEqual(2);
+});
+
+test("beforeResolution multiple interceptors get called correctly", () => {
+  class Bar {}
+  let interceptor1Called = false;
+  let interceptor2Called = false;
+  globalContainer.beforeResolution(
+    Bar,
+    () => {
+      interceptor1Called = true;
+    },
```

Comment:
You could use `jest.fn()` for these, then you can assert `expect(interceptor1).toHaveBeenCalled()` etc

Reply:
I believe `jest.fn()` allows you to assert on the params and call count also

Reply:
Switched this. Much better now :)